warehouse/organizations/models.py

```diff
@@ -0,0 +1,263 @@
+        ForeignKey("projects.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+
+    organization = orm.relationship("Organization", lazy=False)
+    project = orm.relationship("Project", lazy=False)
+
+
+class OrganizationType(enum.Enum):
+
+    Community = "Community"
+    Company = "Company"
+
+
+class OrganizationFactory:
+    def __init__(self, request):
+        self.request = request
+
+    def __getitem__(self, organization):
+        try:
+            return (
+                self.request.db.query(Organization)
+                .filter(
+                    Organization.normalized_name
+                    == func.normalize_pep426_name(organization)
+                )
+                .one()
+            )
+        except NoResultFound:
+            raise KeyError from None
+
+
+# TODO: Determine if this should also utilize SitemapMixin and TwoFactorRequireable
+# class Project(SitemapMixin, TwoFactorRequireable, db.Model):
+class Organization(db.Model):
+    __tablename__ = "organizations"
+    __table_args__ = (
+        CheckConstraint(
+            "name ~* '^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$'::text",
+            name="organizations_valid_name",
+        ),
+    )
+
+    __repr__ = make_repr("name")
+
+    name = Column(Text, nullable=False)
+    normalized_name = orm.column_property(func.normalize_pep426_name(name))
+    display_name = Column(Text, nullable=False)
+    orgtype = Column(
+        Enum(OrganizationType, values_callable=lambda x: [e.value for e in x]),
+        nullable=False,
+    )
+    link_url = Column(URLType, nullable=False)
+    description = Column(Text, nullable=False)
+    is_active = Column(Boolean, nullable=False, default=False)
+    is_approved = Column(Boolean)
+    created = Column(
+        DateTime(timezone=False),
+        nullable=False,
+        server_default=sql.func.now(),
+        index=True,
+    )
+    date_approved = Column(
+        DateTime(timezone=False),
+        nullable=True,
+        onupdate=func.now(),
+    )
+
+    # TODO: Determine if cascade applies to any of these relationships
+    users = orm.relationship(
+        User, secondary=OrganizationRole.__table__, backref="organizations"  # type: ignore # noqa
+    )
+    projects = orm.relationship(
+        "Project", secondary=OrganizationProject.__table__, backref="organizations"  # type: ignore # noqa
+    )
+
+    events = orm.relationship(
+        "OrganizationEvent",
+        backref="organizations",
+        cascade="all, delete-orphan",
+        lazy=True,
+    )
+
+    # TODO:
+    #    def __acl__(self):
+
+    def record_event(self, *, tag, ip_address, additional):
+        session = orm.object_session(self)
+        organization = OrganizationEvent(
+            user=self, tag=tag, ip_address=ip_address, additional=additional
+        )
+        session.add(organization)
+        session.flush()
+
+        return organization
+
+
+class OrganizationEvent(db.Model):
```

Comment:
We may want to relate a (nullable) User object to each action *as well* as shoveling some user info into additional data.

That will just make presentation easier in an Org activity view, and allow for User accounts to be deleted. This doesn't need to happen here/now.

Reply:
Switched to using @di’s `Organization.Event` model from #8324.